### PR TITLE
Bugfix/slon

### PR DESCRIPTION
**Pull request by godsic**
mumax3 assumes the fixed layer is placed on top of the free one, which is consistent with OOMMF implementation. However there are at least two issues with this notation:

- most of the mumax3 people are not aware of this assumption (although it is clearly stated in the mumax3 paper) and use Slonczewski torque blindly;
- most of the real-world samples are made in the opposite way, i.e. fixed layer underneath the free one, similar to how Slonczewski derived his original equation.

So in mumax3 if one want to switch parallel layer a _negative_ current has to be applied, while in Slonczewski paper a _positive_ current is needed. 

To avoid this confusion, I propose to introduce a FixedLayerPosition quantity that might have one of the following values:
FIXEDLAYER_TOP (default) which does not change how mumax3 behaves, i.e. one needs negative current to switch parallel layers;
FIXEDLAYER_BOTTOM which flips the signs of both damping-like and field-like terms, so that positive current is needed to switch parallel layers.

### CODE REVIEW

File: engine/torque.go [hunk 1]
```diff
@@ -33,6 +34,9 @@ func init() {
 	DeclVar("DisableZhangLiTorque", &DisableZhangLiTorque, "Disables Zhang-Li torque (default=false)")
 	DeclVar("DisableSlonczewskiTorque", &DisableSlonczewskiTorque, "Disables Slonczewski torque (default=false)")
 	DeclVar("DoPrecess", &Precess, "Enables LL precession (default=true)")
+	DeclLValue("FixedLayerPosition", &flposition{}, "Position of the fixed layer: FIXEDLAYER_TOP, FIXEDLAYER_BOTTOM (default=FIXEDLAYER_TOP)")
+	DeclROnly("FIXEDLAYER_TOP", FIXEDLAYER_TOP, "FixedLayerPosition = FIXEDLAYER_TOP instructs mumax3 that fixed layer is on top of the free layer")
+	DeclROnly("FIXEDLAYER_BOTTOM", FIXEDLAYER_BOTTOM, "OutputFormat = OVF2_BINARY instructs mumax3 that fixed layer is underneath the free layer")
```
Comment (barnex) [inline engine/torque.go]: copy-pasta typo: OutputFormat = OVF2_BINARY

File: engine/torque.go [hunk 2]
```diff
@@ -98,6 +105,7 @@ func AddSTTorque(dst *data.Slice) {
 		defer epsPrime.Recycle()
 		cuda.AddSlonczewskiTorque2(dst, M.Buffer(),
 			msat, j, fixedP, alpha, pol, lambda, epsPrime, Mesh())
+		j.SetMul(Z, jmul_z)
```
Comment (barnex) [inline engine/torque.go]: redundant because of previous defer j.SetMul... ?

**Comment by godsic** [comment]
@barnex thanks for the feedback. It turned out that if current is set from file, the multiplier got overridden when cuda kernel in invoked. So I have re-worked this pull request to solve this issues. Please check it out. There will be a separate pull request to fix the sign of the pre-prefactor in the Slonczewski damping-like torque.

**Comment by godsic** [comment]
before 53dbec4 there was a significant discrepancy between OOMMF and mumax3 when un-physically large value of the non-adiabaticity is used (as set by the secondary spin-torque parameter):

before-fix

As it has been pointed out in #99 the problem is in the pre-factor of the damping-like term. So 53dbec4 flips the sign as suggested in #99 so that we get a perfect agreement with OOMMF:

fix

I would not expect any major harm from this issue, since with realistic values of the secondary spin-torque parameter the effect is at least of the second order.

**Pull request merged**